javascript/elements/updates_for_element.js

```diff
@@ -45,71 +45,105 @@ export default class UpdatesForElement extends SubscribingElement {
+          const response = await graciouslyFetch(url, {
+            'X-Cable-Ready': 'update'
+          })
+          this.html[url] = await response.text()
+        }
+      })
+    )
+
     // track current block index for each URL; referred to as fragments
     this.index = {}
 
-    this.blocks.forEach(this.processBlock.bind(this, data))
+    this.blocks.forEach(block => {
+      // if the block's URL is not in the index, initialize it to 0; otherwise, increment it
+      this.index.hasOwnProperty(block.url)
+        ? this.index[block.url]++
+        : (this.index[block.url] = 0)
+
+      block.process(data, this.html, this.index)
+    })
   }
 
-  async processBlock (data, block) {
-    // memoize the block's URL to avoid unnecessary DOM interop
-    const blockURL = url(block)
+  get blocks () {
+    // memoize blocks to avoid unnecessary DOM traversal
+    if (!this._blocks)
+      this._blocks = Array.from(
+        document.querySelectorAll(this.query),
+        element => new Block(element)
+      )
 
-    // if the block's URL is not in the index, initialize it to 0; otherwise, increment it
-    this.index.hasOwnProperty(blockURL)
-      ? this.index[blockURL]++
-      : (this.index[blockURL] = 0)
+    return this._blocks
+  }
 
-    // with the index incremented, we can now safely bail - before a fetch - if there's no work to be done
-    if (!this.shouldUpdate(data, block)) return
+  get query () {
+    return `updates-for[identifier="${this.identifier}"]`
+  }
 
-    const index = this.index[blockURL]
+  get identifier () {
+    return this.getAttribute('identifier')
+  }
 
-    // we only want to fetch each URL once
-    if (!this.html.hasOwnProperty(blockURL)) {
-      const response = await graciouslyFetch(blockURL, {
-        'X-Cable-Ready': 'update'
-      })
-      this.html[blockURL] = await response.text()
-    }
+  get debounce () {
+    return this.hasAttribute('debounce')
+      ? parseInt(this.getAttribute('debounce'))
+      : 20
+  }
+}
+
+class Block {
+  constructor (element) {
+    this.element = element
+  }
 
-    block.setAttribute('updating', 'updating')
+  async process (data, html, index) {
+    // with the index incremented, we can now safely bail - before a fetch - if there's no work to be done
+    if (!this.shouldUpdate(data)) return
 
+    const blockIndex = index[this.url]
     const template = document.createElement('template')
-    template.innerHTML = String(this.html[blockURL]).trim()
+    this.element.setAttribute('updating', 'updating')
+
+    template.innerHTML = String(html[this.url]).trim()
 
     await this.resolveTurboFrames(template.content)
 
     const fragments = template.content.querySelectorAll(this.query)
 
-    if (fragments.length <= index) {
-      console.warn('Update aborted due to insufficient updates-for elements')
+    if (fragments.length <= blockIndex) {
+      console.warn('Update aborted due to mismatched number of elements')
```

Comment:
I am fine with changing the language, although please consider that this will only catch scenarios where there are not enough blocks to complete the update present. If there are more than expected, they will just be ignored.

I wonder if it makes sense to include the block url in the warning message? It might assist the developer in tracking down which page is problematic.